internal/cache/file/downloader/sparse_downloads_job_test.go

```diff
@@ -0,0 +1,304 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// File that contains tests specific to sparse downloads job.
+
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sync/atomic"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/v3/cfg"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/cache/util"
+	testutil "github.com/googlecloudplatform/gcsfuse/v3/internal/util"
+	. "github.com/jacobsa/ogletest"
+)
+
+// TestSparseDownloader runs all tests for sparse file downloads
+func TestSparseDownloader(t *testing.T) { RunTests(t) }
+
+type sparseDownloaderTest struct {
+	downloaderTest
+}
+
+func init() { RegisterTestSuite(&sparseDownloaderTest{}) }
+
+func (dt *sparseDownloaderTest) SetUp(*TestInfo) {
+	dt.defaultFileCacheConfig = &cfg.FileCacheConfig{
+		EnableSparseFile:    true,
+		DownloadChunkSizeMb: 20, // 20MB chunks for sparse files
+		EnableCrc:           true,
+	}
+	dt.setupHelper()
+}
+
+func (dt *sparseDownloaderTest) Test_calculateSparseChunkBoundaries() {
+	tests := []struct {
+		name           string
+		offset         int64
+		requiredOffset int64
+		chunkSizeMb    int64
+		objectSize     uint64
+		expectedStart  uint64
+		expectedEnd    uint64
+		expectError    bool
+	}{
+		{
+			name:           "single chunk - aligned",
+			offset:         0,
+			requiredOffset: 10 * util.MiB,
+			chunkSizeMb:    20,
+			objectSize:     100 * util.MiB,
+			expectedStart:  0,
+			expectedEnd:    20 * util.MiB,
+			expectError:    false,
+		},
+		{
+			name:           "single chunk - unaligned start",
+			offset:         5 * util.MiB,
+			requiredOffset: 10 * util.MiB,
+			chunkSizeMb:    20,
+			objectSize:     100 * util.MiB,
+			expectedStart:  0, // rounds down to chunk boundary
+			expectedEnd:    20 * util.MiB,
+			expectError:    false,
+		},
+		{
+			name:           "single chunk - unaligned end",
+			offset:         15 * util.MiB,
+			requiredOffset: 25 * util.MiB,
+			chunkSizeMb:    20,
+			objectSize:     100 * util.MiB,
+			expectedStart:  0,
+			expectedEnd:    40 * util.MiB, // rounds up to chunk boundary
+			expectError:    false,
+		},
+		{
+			name:           "chunk end capped at object size",
+			offset:         90 * util.MiB,
+			requiredOffset: 95 * util.MiB,
+			chunkSizeMb:    20,
+			objectSize:     100 * util.MiB,
+			expectedStart:  80 * util.MiB,
+			expectedEnd:    100 * util.MiB, // capped at object size, not 120MB
```

Comment:
with the current logic if the objectSize is 120MB, dont we end up downloading 2 chunks 80-100Mb and 100-120MB when we need only one chunk: 80-100MB.
Can we compute the end chunk as follows:
if requiredOffset%chunkSize > 0 
endChunk =( requiredOffset/chunkSize)+1